### PR TITLE
[WIP] OpenVMS

### PR DESCRIPTION
**Pull request by noci2012**
##### SUMMARY
Work in progress: build Support for OpenVMS, using DCL command across a SSH session.

- recognizing DCL scripts (if they need to land on ansible host), not sure yet.
- adjust SSH to be able to connect to OpenVMS (use_tty issue).
- DCL is a different command language. No "shell" program, it is always resident.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
OpenVMS

##### ADDITIONAL INFORMATION
Still untested, first pointers are needed to some to be expected issues to solve:
* Quoting, DCL doesn't do Quoting like Unix. "........" = String     'variable' = Variable expansion ''variable'  is variable expansion inside a string.   escaping a "in a string = ""    (" with quote ""quote"" ")
* filenames:     consist of REMOTE-SYSTEM::DEVICE:[DIRECTORY.SUB...]FILENAME.TYPE;VERSION
 + Remote system may be of forms: REMOTENODE or REMOTENODE"username" REMOTENODE"username password"   (it may be a logical containing this)
 + DEVICE a disk name or a logical refering to one
 + DIRECTORY a directory tree intersperced with . character delimited with either [ ] or < >.
 + FILENAME, TYPE etc. like other systems, 
 + VERSION a number 1..32767  (-1 = previous version, 0 = current version, -0 is oldest version)
A logical is a kernel based namespace of translations for various objects in the system. 
 + DIRECTORYNAMES, FILENAME & FILETYPE can be 39 characters long on most systems, complete path 255 characters. (All ASCII, Byte sized, uppercase only + $ & _, * and % are used as wildcards). 
 + There is a filesystem that allows Mixed case, and longer elements. ^is used to escape special characters in that space.

Location for temporary files: SYS$SCRATCH: 

Become (SUDO, RUNAS etc. not supported, then use a different logon).

Is Python on the target required? 2.7 is available if needed.

### CODE REVIEW

File: lib/ansible/plugins/connection/ssh.py [hunk 1]
```diff
@@ -260,7 +260,7 @@
             version_added: '2.7'
       use_tty:
         version_added: '2.5'
-        default: 'yes'
+        default: 'yes'  forced to no for OpenVMS
```
Comment (bcoca) [inline lib/ansible/plugins/connection/ssh.py]: dont add comments to default values, this breaks default processing

File: lib/ansible/plugins/shell/dcl.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    #    #    cmd = shlex_quote(cmd)
+    #    #cmd_parts = [env_string.strip(), shebang.replace("#!", "").strip(), cmd]
+    #    #if arg_path is not None:
+    #    #    cmd_parts.append(arg_path)
+    #    #new_cmd = " ".join(cmd_parts)
+    #    new_cmd = cmd
+    #    return new_cmd
+
+    ## needed for? if checksums need to be crateds then until now: CRC/MD5 is possible with native tooling
+    ## I know HPE left something (Maintenance..) out the last few years.
```
Comment (bcoca) [inline lib/ansible/plugins/shell/dcl.py]: this was used for basic file comparison, mostly the stat module now takes care of this

File: lib/ansible/plugins/connection/ssh.py (hunk 1 above)
Comment (noci2012) [reply lib/ansible/plugins/connection/ssh.py]: updated.

File: lib/ansible/plugins/shell/dcl.py (hunk 2 above)
Comment (noci2012) [reply lib/ansible/plugins/shell/dcl.py]: so it can go away.

File: lib/ansible/executor/module_common.py [hunk 3]
```diff
@@ -940,6 +944,16 @@ def _find_module_utils(module_name, b_module_data, module_path, module_args, tas
         )))
         b_module_data = output.getvalue()
 
+ elif module_substyle == 'dcl':
+        # Start of DCL support
+        # shebang needs to start with $!....
+        shebang = u'$!'
```
Comment (bcoca) [inline lib/ansible/executor/module_common.py:950]: or we can move 'shebang' detection to the shell module (might also help with powershell @nitzmahone @jborean93 ?)

File: lib/ansible/plugins/shell/dcl.py (hunk 2 above)
Comment (noci2012) [reply lib/ansible/plugins/shell/dcl.py]: removed.

**Comment by noci2012** [comment]
current fails are problems with two (not three as in the header) test environments.
Appearantly some files could not be downloaded.

**Comment by nitzmahone** [comment]
Hey @noci2012 - we've been kicking around different ideas for pluggable runtime module assembly behaviors for awhile now, and this would definitely fall into that category. 

Without that in place, I'm all for integrating the minor tweaks necessary to make custom modules runnable on OpenVMS, but we'd probably reject explicit core engine support for code sharing on OpenVMS modules, or shipping the connection/shell plugins for it. We're unable to test it, and we've rejected requests in the past for shared code support for other languages (eg perl, ruby). With the new collections support in 2.8 making it easier to ship broader support packages like this outside the Ansible distribution, we'll likely start making the bar for what gets shipped "in the box" much higher. 

Once a couple of the tweaks have been merged, everything but the module replacer/shared-code-assembly support can be done locally without modification to Ansible via custom plugins (and even easier/more transparently with the new collections support in 2.8). You'd just have to duplicate the shared code in each module for now- we'd treat it like a binary module.

Pluggable module assembly will be on the 2.9 roadmap (as it's something that's generally necessary for a lot of different areas), but no promises on availability.

**Comment by noci2012** [comment]
@nitzmahone 
FYI, I am not an experience Python programmer, not deeply acquainted to ansible as such, i do see it could be useful though.

So how should i read your comment like:  Quit doing what i am doing atm as it is a road to nowhere.? [ now is a time i may spend some time on it, that window will pass surely in a few months ].   

I am not sure how the ansible engine can become usable without some work on it.
Biggest problem now is: OpenVMS has no (unix alike) Shell. So how to do discovery, collect data, ....
  quoting is radically different.    command -v... not availabe, show symbol is probably closest to it.
(many commands are available using symbols, aka shell variables more or less like alias).
Then it also ha quite different settings, commands syntax...
Filename:    DEVICE:[dir.subdir]filename.type;version
Protections: S[ysttem]:RWED,O[wner]:RWED,G[roup]:RWED,W[orld]:RWED
chmod 644 somefile   -> SET FILE/PROTECT=(S:RWD,O:RWD,G:R,W:R) somefile
chown john otherfile.x  -> SET FILE/OWNER=JOHN otherfile.x
(or combined: SET FILE/OWNER=JOHN/PROT=S:RWD,O:RWD,G:R,W:R) somefile,otherfile.x )
Files cannot be quoted neither other parameters.

I found out today that multiline input won't cut it using the SSH EXEC interface.
The closest might be a PIPE command which can handle a unix style command line to some extend and with quite some limitations...    pipe command ; command ; command  / pipe command | command | command... >,  < are possible, >/dev/null should become >NL:   unix path's can be used inside perl/python/c programs (with some quirks) but not in DCL commands. 
Here PIPE "could be seen " as a shell, only the -c needs to go.

Quoting needs to be done DCL style, command line length is limited on older version of OpenVMS to 1KB and recent version can handle 4KB. (after expansion of all variables).  (pre V7.0 are even more limited).

wrt. test machines that might be solvable....? I can as about possbilities next Wed. Apr 10 about those.

**Comment by noci2012** [comment]
@nitzmahone
I tried to read up on "Ansible Pluggable Module Assembly"  it just isn't something i can find in the documentation. A google search refers to the Glossary because the words Module, Pluggable & Assembly all are there, just not in one sentence.
Any pointers?

**Comment by bcoca** [comment]
@noci2012 the 'module plugable assembly' is a 'theoretical' term, it refers mostly to the different ways to assemble and package a module for distribution and execution to the target, mostly a generically plugable way to add your dcl replacer code and similar from `mod_args.py` in a user pluggable fashion.  So don't expect there to be any docs, it is mostly a concept that has not crystallized anywhere yet.

I would not be discouraged by that possible roadmap item as even if that does get developed, most of your effort here would still be useful for OpenVMS integration, if that pluggable assembly is completed it should not require too much effort to reuse your replacer code in a new plugin.

As for multiline problems, i'm not sure i can be much help, other shells also lacked pipes but had other ways to chain command outputs.

**Comment by Akasurde** [comment]
Hi @noci2012, Thank you very much for your interest in Ansible. This plugin/module is no longer maintained in this repository and has been migrated to https://github.com/ansible-collections/community.general<br><br>If you have further questions please stop by IRC or the mailing list:<br><br>* IRC: #ansible on irc.freenode.net<br>* mailing list: https://groups.google.com/forum/#!forum/ansible-project<br><br>needs_info<br>

**Pull request closed without merging**